paddle/fluid/operators/detail/rpc_processor.h

```diff
@@ -0,0 +1,172 @@
+  void SetExecutor(framework::Executor* executor) { executor_ = executor; }
+  void SetPrefetchPreparedCtx(
+      std::unique_ptr<framework::ExecutorPrepareContext> prepared) {
+    prefetch_ctx_.reset(prepared.release());
+  }
+
+  // Used for async
+  void SetGradToPreparedCtx(
+      std::unordered_map<
+          std::string, std::shared_ptr<framework::ExecutorPrepareContext>>* g) {
+    grad_to_prepared_ctx_ = g;
+  }
+
+  bool sync_mode() { return sync_mode_; }
+  framework::Scope* scope() { return scope_; }
+  const platform::DeviceContext* dev_ctx() { return dev_ctx_; }
+  framework::ExecutorPrepareContext* prefetch_ctx() {
+    return prefetch_ctx_.get();
+  }
+  framework::ProgramDesc* program() { return program_; }
+  framework::Executor* executor() { return executor_; }
+
+ protected:
+  bool sync_mode_;
+  framework::Scope* scope_;
+  const platform::DeviceContext* dev_ctx_;
+
+  std::unique_ptr<framework::ExecutorPrepareContext> prefetch_ctx_;
+  framework::ProgramDesc* program_;
+  framework::Executor* executor_;
+
+  // Used for async.
+  std::unordered_map<std::string,
+                     std::shared_ptr<framework::ExecutorPrepareContext>>*
+      grad_to_prepared_ctx_;
+};
+
+class GRPCProcessorCtx : public RPCProcessorCtx {
+ public:
+  GRPCProcessorCtx() : exit_flag_(false), fan_in_(-1) { clear_to_init(); }
+  virtual ~GRPCProcessorCtx() {}
+
+  bool ProcessSendImpl(const std::string& msg_name, framework::Variable* var,
+                       framework::Scope* scope = nullptr);
+
+  bool ProcessGetImpl(const std::string& msg_name, framework::Variable** var);
+
+  bool ProcessPrefetchImpl(const std::string& msg_name, framework::Scope* scope,
+                           framework::Variable** var);
+
+  void SetFanIn(int fan_in) { fan_in_ = fan_in; }
+
+  void WaitFanInOfSend() {
+    std::unique_lock<std::mutex> lock(this->mutex_);
+    condition_send_.wait(lock, [=] {
+      return (this->batch_barrier_send_ >= fan_in_ || exit_flag_);
+    });
+
+    VLOG(3) << "batch_barrier_send_:" << batch_barrier_send_;
+  }
+
+  void WaitFanInOfGet() {
+    std::unique_lock<std::mutex> lock(this->mutex_);
+    condition_get_.wait(lock, [=] {
+      return (this->batch_barrier_get_ >= fan_in_ || exit_flag_);
+    });
+
+    VLOG(3) << "batch_barrier_get_:" << batch_barrier_get_;
+  }
+
+  void clear_to_init() {
+    std::unique_lock<std::mutex> lock(this->mutex_);
+    batch_barrier_send_ = 0;
+
+    sparse_vars_.clear();
+    batch_barrier_get_ = 0;
+  }
+
+  void SetExit();
+
+  bool IsExit() {
+    std::unique_lock<std::mutex> lock(this->mutex_);
+    return exit_flag_;
+  }
+
+  std::vector<framework::Variable*>& sparse_vars() { return sparse_vars_; }
+
+ private:
+  void IncreaseBatchBarrierGet();
+  void IncreaseBatchBarrierSend();
+  // void IncreaseRecvVarCnt();
+
+ private:
+  // status
+  bool exit_flag_;
+  int fan_in_;
+  std::mutex mutex_;
+
+  // send
+  std::condition_variable condition_send_;
```

Comment:
Only one condition is needed here for all status. The "condition" is the current "state" of the RCP server, condition will be changed only when client barrier reaches.

Reply:
Done.